rpcs3/Emu/RSX/Common/BufferUtils.cpp

```diff
@@ -755,36 +756,286 @@ namespace
+			tmp = __mm_shuffle_epi8(min, mask_step2);
+			min = __mm_min_epu32(min, tmp);
+
+			tmp = __mm_shuffle_epi8(max, mask_step1);
+			max = __mm_max_epu32(max, tmp);
+			tmp = __mm_shuffle_epi8(max, mask_step2);
+			max = __mm_max_epu32(max, tmp);
+
+			const u32 min_index = u32(_mm_cvtsi128_si32(min));
+			const u32 max_index = u32(_mm_cvtsi128_si32(max));
+
+			return std::make_tuple(min_index, max_index, count);
+		}
+
 		template<typename T>
 		static
 		std::tuple<T, T, u32> upload_untouched(gsl::span<to_be_t<const T>> src, gsl::span<T> dst, u32 restart_index, bool skip_restart)
 		{
-			T min_index = index_limit<T>(), max_index = 0;
-			u32 dst_index = 0;
+			T min_index, max_index;
+			u32 written;
+			u32 remaining = src.size();
+
+#if __AVX2__
+			if (s_use_avx2 && remaining >= 64 && !skip_restart)
+			{
+				if constexpr (std::is_same<T, u32>::value)
+				{
+					const auto count = (remaining & ~0x3);
+					std::tie(min_index, max_index, written) = upload_u32_swapped(src.data(), dst.data(), count, restart_index);
+				}
+				else if constexpr (std::is_same<T, u16>::value)
+				{
+					const auto count = (remaining & ~0xf);
+					std::tie(min_index, max_index, written) = upload_u16_swapped_avx2(src.data(), dst.data(), count, restart_index);
+				}
+				else
+				{
+					fmt::throw_exception("Unreachable" HERE);
+				}
 
-			for (const T index : src)
+				remaining -= written;
+			}
+			else
+#endif
+			if (s_use_sse4_1 && remaining >= 32 && !skip_restart)
 			{
+				if constexpr (std::is_same<T, u32>::value)
+				{
+					const auto count = (remaining & ~0x3);
+					std::tie(min_index, max_index, written) = upload_u32_swapped(src.data(), dst.data(), count, restart_index);
+				}
+				else if constexpr (std::is_same<T, u16>::value)
+				{
+					const auto count = (remaining & ~0x7);
+					std::tie(min_index, max_index, written) = upload_u16_swapped_sse4_1(src.data(), dst.data(), count, restart_index);
+				}
+				else
+				{
+					fmt::throw_exception("Unreachable" HERE);
+				}
+
+				remaining -= written;
+			}
+			else
+			{
+				min_index = index_limit<T>();
+				max_index = 0;
+				written = 0;
+			}
+
+			while (remaining--)
+			{
+				T index = src[written];
 				if (index == restart_index)
 				{
 					if (!skip_restart)
 					{
-						dst[dst_index++] = index_limit<T>();
+						dst[written++] = index_limit<T>();
 					}
 				}
 				else
 				{
-					dst[dst_index++] = min_max(min_index, max_index, index);
+					dst[written++] = min_max(min_index, max_index, index);
 				}
 			}
 
-			return std::make_tuple(min_index, max_index, dst_index);
+			return std::make_tuple(min_index, max_index, written);
 		}
 	};
 
 	template<typename T>
 	std::tuple<T, T, u32> upload_untouched(gsl::span<to_be_t<const T>> src, gsl::span<T> dst, rsx::primitive_type draw_mode, bool is_primitive_restart_enabled, u32 primitive_restart_index)
 	{
-		if (LIKELY(!is_primitive_restart_enabled))
+		if (UNLIKELY(!is_primitive_restart_enabled))
```

Comment:
Why is this change?

Reply:
Oh, it was to get the generated assembly more readable by moving the relevant part closer to the top of the function. I have actually no idea which of the two functions is the most common in games.

Reply:
I have now removed this `UNLIKELY`.